x-pack/solutions/security/plugins/cloud_security_posture/server/routes/graph/fetch_graph.ts

```diff
@@ -296,27 +290,59 @@ ${
   sourceIps = MV_DEDUPE(VALUES(sourceIps)),
   sourceCountryCodes = MV_DEDUPE(VALUES(sourceCountryCodes)),
   // actor attributes
-  actorEntityGroup = VALUES(actorEntityGroup),
   actorIds = VALUES(actor.entity.id),
   actorIdsCount = COUNT_DISTINCT(actor.entity.id),
+  actorEntityName = VALUES(actorEntityName),
   actorEntityType = VALUES(actorEntityType),
-  actorLabel = VALUES(actorLabel),
+  actorEntitySubType = VALUES(actorEntitySubType),
   actorsDocData = VALUES(actorDocData),
   actorHostIp = VALUES(actorHostIp),
   // target attributes
-  targetEntityGroup = VALUES(targetEntityGroup),
   targetIds = VALUES(target.entity.id),
   targetIdsCount = COUNT_DISTINCT(target.entity.id),
+  targetEntityName = VALUES(targetEntityName),
   targetEntityType = VALUES(targetEntityType),
-  targetLabel = VALUES(targetLabel),
+  targetEntitySubType = VALUES(targetEntitySubType),
   targetsDocData = VALUES(targetDocData)
     BY action = event.action,
       actorEntityGroup,
       targetEntityGroup,
       isOrigin,
       isOriginAlert
+| EVAL actorEntityType = CASE(
+    actorEntityType IS NOT NULL,
+    actorEntityType,
+    actorIdsCount == 1,
+    "${NON_ENRICHED_ENTITY_TYPE_SINGULAR}",
+    "${NON_ENRICHED_ENTITY_TYPE_PLURAL}"
+  )
+| EVAL targetEntityType = CASE(
+    targetEntityType IS NOT NULL,
+    targetEntityType,
+    targetIdsCount == 1,
+    "${NON_ENRICHED_ENTITY_TYPE_SINGULAR}",
+    "${NON_ENRICHED_ENTITY_TYPE_PLURAL}"
+  )
+| EVAL actorLabel = CASE(
+    actorIdsCount == 1 AND actorEntityType == "${NON_ENRICHED_ENTITY_TYPE_SINGULAR}",
+    TO_STRING(actorIds),
+    actorIdsCount == 1 AND actorEntityType != "${NON_ENRICHED_ENTITY_TYPE_SINGULAR}",
+    actorEntityName,
+    actorIdsCount > 1 AND actorEntitySubType IS NOT NULL,
+    actorEntitySubType,
+    ""
+  )
+| EVAL targetLabel = CASE(
```

Comment:
I highly recommend to keep the ESQL query lean from BL like determine the label

It would be much simpler to return all relevant data like entity's type, sub type and name.
And later determine in the code what should be the label

This way you'd have less repetitive code in the ESQL and much more straight forward tests for that specific logic.